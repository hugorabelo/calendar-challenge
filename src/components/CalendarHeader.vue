<template>
  <div class="header flex items-center p-4">
    <button
      id="btn-today"
      class="border rounded-md h-10 border-color-gray-200 hover:bg-gray-200 p-2 mr-6"
      @click="goToToday()"
    >
      Today
    </button>
    <button id="btn-previous" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-1" @click="goToPreviousMonth()">
      <font-awesome-icon icon="fa-solid fa-chevron-left" />
    </button>
    <button id="btn-next" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-8" @click="goToNextMonth()">
      <font-awesome-icon icon="fa-solid fa-chevron-right" />
    </button>
    <div class="text-3xl md:text-3xl hidden md:flex long-title">
      {{ headerLongTitle }}
    </div>
    <div class="text-lg md:text-3xl md:hidden short-title">
      {{ headerShortTitle }}
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
  name: 'CalendarHeader',
  computed: {
    ...mapGetters(['selectedMonth']),
    headerLongTitle() {
      return this.selectedMonth.toLocaleString({ month: 'long', year: 'numeric' })
    },
    headerShortTitle() {
      return this.selectedMonth.toLocaleString({ month: 'short', year: 'numeric' })
    },
  },
  methods: {
    goToToday() {
      this.$emit('today')
    },
    goToPreviousMonth() {
      this.$emit('previousMonth')
    },
    goToNextMonth() {
      this.$emit('nextMonth')
    },
  },
}
</script>

<style></style>
