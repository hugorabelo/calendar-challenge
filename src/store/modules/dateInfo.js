import { DateTime } from 'luxon'

const state = {
  selectedMonth: DateTime.now().startOf('month'),
}
const getters = {
  selectedMonth: (state) => state.selectedMonth,
}
const mutations = {
  setSelectedMonth(state, data) {
    state.selectedMonth = data
  },
}
const actions = {
  initSelectedMonth({ commit }) {
    commit('setSelectedMonth', DateTime.now().startOf('month'))
  },
  goToNextMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let nextMonth = state.selectedMonth.plus({ months: 1 })
    commit('setSelectedMonth', nextMonth)
  },
  goToPreviousMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let previousMonth = state.selectedMonth.minus({ months: 1 })
    commit('setSelectedMonth', previousMonth)
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
