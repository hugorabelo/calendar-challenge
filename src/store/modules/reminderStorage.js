const state = {
  reminders: localStorage.getItem('reminders') != null ? JSON.parse(localStorage.getItem('reminders')) : [],
}
const getters = {
  reminders: (state) => state.reminders,
}
const mutations = {
  setReminders(state, data) {
    state.reminders = data
    localStorage.setItem('reminders', JSON.stringify(state.reminders))
  },
}
const actions = {
  addReminder({ commit, state }, data) {
    if (state.reminders.filter((rem) => rem.id == data.id).length == 0) {
      state.reminders.push(data)
    }
    commit('setReminders', state.reminders)
  },
  updateReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders[i] = data
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders.splice(i, 1)
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteRemindersGroup({ dispatch }, data) {
    for (var i = 0; i < data.length; i++) {
      dispatch('deleteReminder', { id: data[i].id })
    }
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
