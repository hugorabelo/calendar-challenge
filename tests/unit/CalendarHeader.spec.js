import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import CalendarHeader from '@/components/CalendarHeader'
import { DateTime } from 'luxon'
import dateInfo from '@/store/modules/dateInfo'
import reminderStorage from '@/store/modules/reminderStorage'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faChevronLeft, faChevronRight, faPlus } from '@fortawesome/free-solid-svg-icons'

const localVue = createLocalVue()
localVue.use(Vuex)
library.add(faChevronLeft, faChevronRight, faPlus)
localVue.component('font-awesome-icon', FontAwesomeIcon)

describe('CalendarHeader', () => {
  const build = (options) => {
    return mount(CalendarHeader, {
      localVue,
      store: new Vuex.Store({
        modules: {
          dateInfo: { ...dateInfo },
          reminderStorage: { ...reminderStorage },
        },
      }),
      ...options,
    })
  }

  test('Is a Vue instance', () => {
    const { vm } = build()
    expect(vm).toBeTruthy()
  })

  test('Header title is working', () => {
    const wrapper = build()

    let currentMonth = DateTime.now().startOf('month')
    let longTitle = currentMonth.toLocaleString({ month: 'long', year: 'numeric' })
    let shortTitle = currentMonth.toLocaleString({ month: 'short', year: 'numeric' })

    expect(wrapper.vm.headerLongTitle).toBe(longTitle)

    expect(wrapper.vm.headerShortTitle).toBe(shortTitle)

    expect(wrapper.html()).toMatchSnapshot()

    const longLabel = wrapper.findComponent('.long-title')
    expect(longLabel.text()).toBe(longTitle)

    const shortLabel = wrapper.findComponent('.short-title')
    expect(shortLabel.text()).toBe(shortTitle)
  })

  test('Click buttons', async () => {
    const wrapper = build()

    expect(wrapper.html()).toMatchSnapshot()

    let btnToday = wrapper.findComponent('#btn-today')
    let btnPrevious = wrapper.findComponent('#btn-previous')
    let btnNext = wrapper.findComponent('#btn-next')

    await btnToday.trigger('click')
    expect(wrapper.emitted('today')).toBeTruthy()

    await btnPrevious.trigger('click')
    expect(wrapper.emitted('previousMonth')).toBeTruthy()

    await btnNext.trigger('click')
    expect(wrapper.emitted('nextMonth')).toBeTruthy()
  })
})
