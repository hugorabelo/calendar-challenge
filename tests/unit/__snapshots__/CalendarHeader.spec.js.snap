// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader Click buttons 1`] = `
<div class="header flex items-center p-4"><button id="btn-today" class="border rounded-md h-10 border-color-gray-200 hover:bg-gray-200 p-2 mr-6">
    Today
  </button> <button id="btn-previous" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-1"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-left" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="svg-inline--fa fa-chevron-left">
      <path fill="currentColor" d="M224 480c-8.188 0-16.38-3.125-22.62-9.375l-192-192c-12.5-12.5-12.5-32.75 0-45.25l192-192c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25L77.25 256l169.4 169.4c12.5 12.5 12.5 32.75 0 45.25C240.4 476.9 232.2 480 224 480z" class=""></path>
    </svg></button> <button id="btn-next" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-8"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-right" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="svg-inline--fa fa-chevron-right">
      <path fill="currentColor" d="M96 480c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L242.8 256L73.38 86.63c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l192 192c12.5 12.5 12.5 32.75 0 45.25l-192 192C112.4 476.9 104.2 480 96 480z" class=""></path>
    </svg></button>
  <div class="text-3xl md:text-3xl hidden md:flex long-title">
    março de 2022
  </div>
  <div class="text-lg md:text-3xl md:hidden short-title">
    mar. de 2022
  </div>
</div>
`;

exports[`CalendarHeader Header title is working 1`] = `
<div class="header flex items-center p-4"><button id="btn-today" class="border rounded-md h-10 border-color-gray-200 hover:bg-gray-200 p-2 mr-6">
    Today
  </button> <button id="btn-previous" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-1"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-left" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="svg-inline--fa fa-chevron-left">
      <path fill="currentColor" d="M224 480c-8.188 0-16.38-3.125-22.62-9.375l-192-192c-12.5-12.5-12.5-32.75 0-45.25l192-192c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25L77.25 256l169.4 169.4c12.5 12.5 12.5 32.75 0 45.25C240.4 476.9 232.2 480 224 480z" class=""></path>
    </svg></button> <button id="btn-next" class="rounded-full h-10 w-10 hover:bg-gray-200 mr-8"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-right" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="svg-inline--fa fa-chevron-right">
      <path fill="currentColor" d="M96 480c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L242.8 256L73.38 86.63c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l192 192c12.5 12.5 12.5 32.75 0 45.25l-192 192C112.4 476.9 104.2 480 96 480z" class=""></path>
    </svg></button>
  <div class="text-3xl md:text-3xl hidden md:flex long-title">
    março de 2022
  </div>
  <div class="text-lg md:text-3xl md:hidden short-title">
    mar. de 2022
  </div>
</div>
`;
